### PR TITLE
feat: tasks used in courses

### PR DESCRIPTION
**Pull request by ThorsAngerVaNeT**
**🟢 Add `deploy` label if you want to deploy this Pull Request to staging environment**

#### 🧑‍⚖️ Pull Request Naming Convention

- Title should follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- Do not put issue id in title
- Do not put WIP in title. Use [Draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/) functionality
- Consider to add `area:*` label(s)

* [x] I followed naming convention rules

---

#### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Performance optimization
- [ ] Refactoring
- [ ] Test Case
- [ ] Documentation update
- [ ] Other

#### 🔗 Related issue link

https://github.com/rolling-scopes/rsschool-app/issues/2106

#### 💡 Background and solution

Add column `Used in Courses` with filter to Tasks table and Edit form
<details>

<summary>Screenshots</summary>

image
image
image

</details>

#### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Database migration is added or not needed
- [x] Documentation is updated/provided or not needed
- [x] Changes are tested locally

### CODE REVIEW

File: client/src/components/Table/renderers.tsx [hunk 1]
```diff
@@ -98,6 +98,21 @@ export function tagsRenderer(values: (number | string)[]) {
   return <span>{values.map(v => renderTag(v))}</span>;
 }
 
+export function tagsCoursesRendererWithRemainingNumber(_: undefined, { courses }: TaskDto) {
+  if (!Array.isArray(courses) || courses.length === 0) {
+    return '';
+  }
```
Comment (valerydluski) [inline client/src/components/Table/renderers.tsx:104]:
```
    if (!courses?.length) {
        return '';
    }
```

File: client/src/components/Table/renderers.tsx [hunk 2]
```diff
@@ -98,6 +98,21 @@ export function tagsRenderer(values: (number | string)[]) {
...
+
+  const tags = courses.slice(0, 1).map(({ name, isActive }) => ({ value: name, ...(isActive && { color: 'blue' }) }));
+  const count = courses.length - 1;
+
+  if (courses.length && count) {
```
Comment (valerydluski) [inline client/src/components/Table/renderers.tsx:109]:
```
    const coursesCount = courses.length - 1;

    if (coursesCount > 0) {
```
or
`if (courses.length - 1 > 0) {`

File: client/src/components/Table/renderers.tsx (hunk 1 above)
Comment (ThorsAngerVaNeT) [reply client/src/components/Table/renderers.tsx]: fixed

File: client/src/components/Table/renderers.tsx (hunk 2 above)
Comment (ThorsAngerVaNeT) [reply client/src/components/Table/renderers.tsx]: fixed

File: nestjs/src/courses/dto/used-course.dto.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { IsString } from 'class-validator';
+
+export class UsedCourseDto {
+  constructor(obj: { name: string; isActive: boolean }) {
```
Comment (valerydluski) [inline nestjs/src/courses/dto/used-course.dto.ts:5]: Maybe it's better to use a more understandable naming, like 'course' instead of 'obj'?

Comment (ThorsAngerVaNeT) [reply nestjs/src/courses/dto/used-course.dto.ts]: You're right. Fixed

File: client/src/pages/admin/tasks.tsx [hunk 4]
```diff
@@ -108,6 +130,10 @@ function Page(props: Props) {
     [modalData, modalAction, modalLoading, dataCriteria],
   );
 
+  const allUsedCourses = uniqBy(data.map(({ courses }) => courses).flat(), course => course.name)
```
Comment (valerydluski) [inline client/src/pages/admin/tasks.tsx:133]: can we use flatMap here?

**Review by valerydluski — APPROVED** [review]

File: client/src/pages/admin/tasks.tsx (hunk 4 above)
Comment (ThorsAngerVaNeT) [reply client/src/pages/admin/tasks.tsx]: Good catch! Indeed `flatMap` is perfect here.
Fixed

**Pull request merged**